skada/utils.py

```diff
@@ -0,0 +1,264 @@
+        Allow automatic generation of sample_domain if not provided.
+
+    Returns:
+    ----------
+    X : array
+        Combined input features for source and target domains.
+    sample_domain : array
+        Combined domain labels for source and target domains.
+    """
+    X = check_array(X, input_name='X')
+
+    if sample_domain is None and not allow_auto_sample_domain:
+        raise ValueError("Either 'sample_domain' or 'allow_auto_sample_domain' "
+                         "should be set")
+    elif sample_domain is None and allow_auto_sample_domain:
+        # default target domain when sample_domain is not given
+        # xxx(okachaiev): I guess this should be -inf instead of a number
+        # The idea is that with no labels we always assume
+        # target domain (_DEFAULT_TARGET_DOMAIN_ONLY_LABEL)
+        sample_domain = (
+            _DEFAULT_TARGET_DOMAIN_ONLY_LABEL * np.ones(X.shape[0], dtype=np.int32)
+        )
+
+    source_idx = extract_source_indices(sample_domain)
+    check_consistent_length(X, sample_domain)
+
+    if allow_domains is not None:
+        for domain in np.unique(sample_domain):
+            # xxx(okachaiev): re-definition of the wildcards
+            wildcard = np.inf if domain >= 0 else -np.inf
+            if domain not in allow_domains and wildcard not in allow_domains:
+                raise ValueError(f"Unknown domain label '{domain}' given")
+
+    n_sources = np.unique(sample_domain[source_idx]).shape[0]
+    n_targets = np.unique(sample_domain[~source_idx]).shape[0]
+
+    if not allow_source and n_sources > 0:
+        raise ValueError(f"Number of sources provided is {n_sources} "
+                         "and 'allow_source' is set to False")
+    if not allow_target and n_targets > 0:
+        raise ValueError(f"Number of targets provided is {n_targets} "
+                         "and 'allow_target' is set to False")
+    if not allow_multi_source and n_sources > 1:
+        raise ValueError(f"Number of sources provided is {n_sources} "
+                         "and 'allow_multi_source' is set to False")
+    if not allow_multi_target and n_sources > 1:
+        raise ValueError(f"Number of targets provided is {n_targets} "
+                         "and 'allow_multi_target' is set to False")
+
+    return X, sample_domain
+
+
+def extract_source_indices(sample_domain):
+    """Extract the indices of the source samples.
+
+    Parameters:
+    ----------
+    sample_domain : array-like of shape (n_samples,)
+        Array specifying the domain labels for each sample.
+
+    Returns:
+    ----------
+    source_idx : array
+        Boolean array indicating source indices.
+    """
+    sample_domain = check_array(
+        sample_domain,
+        dtype=np.int32,
+        ensure_2d=False,
+        input_name='sample_domain'
+    )
+
+    source_idx = (sample_domain >= 0)
+    return source_idx
+
+
+def split_source_target_X_y(X, y, sample_domain):
```

Comment:
Rather than having separate functions for `split_source_target_X_y` and `split_source_target_X`, it might be simpler to have a single function `split_source_target` following the call convention of `train_test_split`: the function returns pair for each input given. Something like:

```python
X_source, X_target = split_source_target(X, sample_domain=sample_domain)
X_source, X_target, y_source, y_target = split_source_target(X, y, sample_domain=sample_domain)
X_source, X_target, y_source, y_target, D_source, D_target  = split_source_target(X, y, D, sample_domain=sample_domain)
# ... etc
```

Also, maybe `source_target_split` (similar to `train_test_split` in sklearn)?

Reply:
I agree. Just, there is already a `source_target_split` function in the _utils.py file. Should I move it to utils.py and modify it so he can handle more than just (X, y) as a split input?

Reply:
If we have a new function available from utils we don't need an old one